hal/explicit_pinmap.h

```diff
@@ -0,0 +1,186 @@
+}
+
+MSTD_CONSTEXPR_FN_14 serial_pinmap_t get_uart_pinmap(const PinName tx, const PinName rx)
+{
+    const PinMap *tx_map = nullptr;
+    for (const PinMap &pinmap : PINMAP_UART_TX) {
+        if (pinmap.pin == tx) {
+            tx_map = &pinmap;
+            break;
+        }
+    }
+
+    const PinMap *rx_map = nullptr;
+    for (const PinMap &pinmap : PINMAP_UART_RX) {
+        if (pinmap.pin == rx) {
+            rx_map = &pinmap;
+            break;
+        }
+    }
+
+    if (tx_map->pin == NC || rx_map->pin == NC || rx_map->peripheral != tx_map->peripheral) {
+        return {NC, NC, NC, NC, NC, 0};
+    }
+
+    if (tx_map->pin == STDIO_UART_TX && rx_map->pin == STDIO_UART_RX) {
+        return {tx_map->peripheral, tx_map->pin, tx_map->function, rx_map->pin, rx_map->function, 1};
+    } else {
+        return {tx_map->peripheral, tx_map->pin, tx_map->function, rx_map->pin, rx_map->function, 0};
+    }
+}
+
+MSTD_CONSTEXPR_FN_14 serial_fc_pinmap_t get_uart_fc_pinmap(const PinName rxflow, const PinName txflow)
+{
+    const PinMap *rts_map = nullptr;
+    for (const PinMap &pinmap : PINMAP_UART_RTS) {
+        if (pinmap.pin == rxflow) {
+            rts_map = &pinmap;
+            break;
+        }
+    }
+
+    const PinMap *cts_map = nullptr;
+    for (const PinMap &pinmap : PINMAP_UART_CTS) {
+        if (pinmap.pin == txflow) {
+            cts_map = &pinmap;
+            break;
+        }
+    }
+
+    if (((rts_map->pin != NC && cts_map->pin != NC) && rts_map->peripheral != cts_map->peripheral) || (rts_map->pin == NC && cts_map->pin == NC)) {
+        return {NC, NC, NC, NC, NC};
+    }
+
+    return {cts_map->peripheral, cts_map->pin, cts_map->function, rts_map->pin, rts_map->function};
+}
+
+MSTD_CONSTEXPR_FN_14 spi_pinmap_t get_spi_pinmap(const PinName mosi, const PinName miso, const PinName sclk, const PinName ssel)
+{
+    const PinMap *mosi_map = nullptr;
+    for (const PinMap &pinmap : PINMAP_SPI_MOSI) {
+        if (pinmap.pin == mosi) {
+            mosi_map = &pinmap;
+            break;
+        }
+    }
+
+    const PinMap *miso_map = nullptr;
+    for (const PinMap &pinmap : PINMAP_SPI_MISO) {
+        if (pinmap.pin == miso) {
+            miso_map = &pinmap;
+            break;
+        }
+    }
+
+    const PinMap *sclk_map = nullptr;
+    for (const PinMap &pinmap : PINMAP_SPI_SCLK) {
+        if (pinmap.pin == sclk) {
+            sclk_map = &pinmap;
+            break;
+        }
+    }
+
+    const PinMap *ssel_map = nullptr;
+    for (const PinMap &pinmap : PINMAP_SPI_SSEL) {
+        if (pinmap.pin == ssel) {
+            ssel_map = &pinmap;
+            break;
+        }
+    }
+
+    if ((mosi_map->pin == NC || miso_map->pin == NC || sclk_map->pin == NC) ||
+        mosi_map->peripheral != miso_map->peripheral ||
+        mosi_map->peripheral != sclk_map->peripheral ||
+    (ssel_map->pin != NC && mosi_map->peripheral != ssel_map->peripheral)) {
```

Comment:
This clause should presumably be `ssel != NC && (!ssel_map ||  mosi_map->peripheral != ssel_map->peripheral)`

Reply:
Each table ends with `{NC   , NC    , 0}` row, if given `ssel == NC`, then the last row should be selected, so we can use `ssel_map->pin` which should be `NC`.

Reply:
But still the entire `if` statement needs to be corrected.

Reply:
But if `ssel` was invalid, then `ssel_map` is nullptr, so you need to catch that case.

Reply:
```
    if ((!mosi_map || !miso_map || !sclk_map || !ssel_map) ||
        (mosi_map->peripheral != miso_map->peripheral || mosi_map->peripheral != sclk_map->peripheral) ||
        (ssel_map->pin != NC && mosi_map->peripheral != ssel_map->peripheral)) {
        return {NC, NC, NC, NC, NC, NC, NC, NC, NC};
    }
```

Reply:
Yes, I think that's correct.

Reply:
In such case we will return `{NC, NC, 0, NC, 0, NC, 0}`. This should be also ok.
Don't know why in the last row is `0` for pin function (instead `NC`). I used `NC` since `pinmap_find_function()` returns NC if not found.